hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/RegionScannerImpl.java

```diff
@@ -130,19 +130,34 @@ private static boolean hasNonce(HRegion region, long nonce) {
     long mvccReadPoint = PackagePrivateFieldAccessor.getMvccReadPoint(scan);
     this.scannerReadPoints = region.scannerReadPoints;
     this.rsServices = region.getRegionServerServices();
-    synchronized (scannerReadPoints) {
-      if (mvccReadPoint > 0) {
-        this.readPt = mvccReadPoint;
-      } else if (hasNonce(region, nonce)) {
-        this.readPt = rsServices.getNonceManager().getMvccFromOperationContext(nonceGroup, nonce);
-      } else {
-        this.readPt = region.getReadPoint(isolationLevel);
+    if (region.useReadWriteLockForReadPoints) {
+      region.scannerReadPointsLock.readLock().lock();
+      try {
+        this.readPt = calculateReadPoint(isolationLevel, mvccReadPoint, nonceGroup, nonce);
+        scannerReadPoints.put(this, this.readPt);
```

Comment:
Here we also modified the scannerReadPoints? Why just use read lock here?

Reply:
Thanks for the review Duo

The scannerReadPoints is ConcurrentHashMap, which is thread-safe. So we can modified it concurrently. Shaed lock is enough. But when getSmallestReadPoint, we need to ensure that scannerReadPoints can not be modified, we need exclusive lock.

So when modify it, we need to acquire shared lock and when read it, we need to acquire exclusive lock.

Reply:
Then I think what we want to protect here is not scannerReadPoints itself, but a special operation? So I prefer we change the name of the lock, maybe call it smallestReadPointsCalcLock?

Reply:
ok,  let me rename it and fix the UT problems, Thanks Duo.

Reply:
And why do we need to introduce a new flag here?

Reply:
When we do delta operation like increment/append, we will need to get the smallest read point.

If we both have heavy delta operation load and heavy read load,  the write lock may reduce the write performance. so let's introduce a flag to make this decided by the user.

Reply:
https://github.com/apache/hbase/blob/master/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegion.java#L5239

Reply:
Then I suggest we implement a class called ReadPointCalcuationLock, and introduce two types of lock, one is lockForCalculation, the other is lockForRecording. And we put the flag in this class, when it is enabled, we use a read write lock, and use write lock for calculation, and use read lock for recording. when it is disabled, we just use a lock.

I think this could make the code cleaner. WDYT?

Thanks.

Reply:
I agree. let me address it. 

Thanks for your suggestion. Duo.